plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -256,534 +254,666 @@ struct block_time_tracker {
 
-            if( chain.is_known_unexpired_transaction( id )) {
-               auto except_ptr = std::static_pointer_cast<fc::exception>( std::make_shared<tx_duplicate>(
-                     FC_LOG_MESSAGE( error, "duplicate transaction ${id}", ("id", id))));
-               next( std::move(except_ptr) );
-               return true;
-            }
+         exhausted = pr.block_exhausted;
+         if (pr.trx_exhausted) {
+            _unapplied_transactions.add_incoming(trx, api_trx, return_failure_trace, next);
+         }
 
-            if( !chain.is_building_block()) {
-               _unapplied_transactions.add_incoming( trx, api_trx, return_failure_trace, next );
-               return true;
-            }
+      } catch (const guard_exception& e) {
+         chain_plugin::handle_guard_exception(e);
+      } catch (boost::interprocess::bad_alloc&) {
+         chain_apis::api_base::handle_db_exhaustion();
+      } catch (std::bad_alloc&) {
+         chain_apis::api_base::handle_bad_alloc();
+      }
+      CATCH_AND_CALL(next);
 
-            const auto block_deadline = _pending_block_deadline;
-            push_result pr = push_transaction( block_deadline, trx, api_trx, return_failure_trace, next );
+      return !exhausted;
+   }
 
-            exhausted = pr.block_exhausted;
-            if( pr.trx_exhausted ) {
-               _unapplied_transactions.add_incoming( trx, api_trx, return_failure_trace, next );
-            }
 
-         } catch ( const guard_exception& e ) {
-            chain_plugin::handle_guard_exception(e);
-         } catch ( boost::interprocess::bad_alloc& ) {
-            chain_apis::api_base::handle_db_exhaustion();
-         } catch ( std::bad_alloc& ) {
-            chain_apis::api_base::handle_bad_alloc();
-         } CATCH_AND_CALL(next);
+   fc::microseconds get_irreversible_block_age() {
+      auto now = fc::time_point::now();
+      if (now < _irreversible_block_time) {
+         return fc::microseconds(0);
+      } else {
+         return now - _irreversible_block_time;
+      }
+   }
 
-         return !exhausted;
+   account_name get_pending_block_producer() {
+      auto& chain = chain_plug->chain();
+      if (chain.is_building_block()) {
+         return chain.pending_block_producer();
+      } else {
+         return {};
       }
+   }
 
+   bool production_disabled_by_policy() {
+      return !_production_enabled || _pause_production ||
+             (_max_irreversible_block_age_us.count() >= 0 && get_irreversible_block_age() >= _max_irreversible_block_age_us);
+   }
 
-      fc::microseconds get_irreversible_block_age() {
-         auto now = fc::time_point::now();
-         if (now < _irreversible_block_time) {
-            return fc::microseconds(0);
-         } else {
-            return now - _irreversible_block_time;
-         }
-      }
+   bool is_producer_key(const chain::public_key_type& key) const { return _signature_providers.find(key) != _signature_providers.end(); }
```

Comment:
Hmm, not a fan of these one liners. I prefer one line `if`s over one line methods. But if we are going to force `if` to two lines seems like we should force these to three lines.

Reply:
fixed